diff -Nur dhcp-4.1.1.orig/client/Makefile.in dhcp-4.1.1/client/Makefile.in
--- dhcp-4.1.1.orig/client/Makefile.in	2012-03-12 19:06:19.440825884 +0800
+++ dhcp-4.1.1/client/Makefile.in	2012-03-12 19:03:08.432523948 +0800
@@ -236,12 +236,12 @@
 
 clean-sbinPROGRAMS:
 	-test -z "$(sbin_PROGRAMS)" || rm -f $(sbin_PROGRAMS)
-#dhclient$(EXEEXT): $(dhclient_OBJECTS) $(dhclient_DEPENDENCIES) 
-#	@rm -f dhclient$(EXEEXT)
-#	$(LINK) $(dhclient_OBJECTS) $(dhclient_LDADD) $(LIBS)
 dhclient$(EXEEXT): $(dhclient_OBJECTS) $(dhclient_DEPENDENCIES) 
 	@rm -f dhclient$(EXEEXT)
-	$(CC) -o $@ $^ $(dhclient_LDADD) $(LIBS)
+	$(LINK) $(dhclient_OBJECTS) $(dhclient_LDADD) $(LIBS)
+#dhclient$(EXEEXT): $(dhclient_OBJECTS) $(dhclient_DEPENDENCIES) 
+#	@rm -f dhclient$(EXEEXT)
+#	$(CC) -o $@ $^ $(dhclient_LDADD) $(LIBS)
 
 mostlyclean-compile:
 	-rm -f *.$(OBJEXT)
diff -Nur dhcp-4.1.1.orig/relay/Makefile.in dhcp-4.1.1/relay/Makefile.in
--- dhcp-4.1.1.orig/relay/Makefile.in	2012-03-12 19:06:19.419832466 +0800
+++ dhcp-4.1.1/relay/Makefile.in	2012-03-12 19:03:08.202327620 +0800
@@ -215,12 +215,12 @@
 
 clean-sbinPROGRAMS:
 	-test -z "$(sbin_PROGRAMS)" || rm -f $(sbin_PROGRAMS)
-#dhcrelay$(EXEEXT): $(dhcrelay_OBJECTS) $(dhcrelay_DEPENDENCIES) 
-#	@rm -f dhcrelay$(EXEEXT)
-#	$(LINK) $(dhcrelay_OBJECTS) $(dhcrelay_LDADD) $(LIBS)
 dhcrelay$(EXEEXT): $(dhcrelay_OBJECTS) $(dhcrelay_DEPENDENCIES) 
 	@rm -f dhcrelay$(EXEEXT)
-	$(CC) -o $@ $^ $(dhcrelay_LDADD) $(LIBS)
+	$(LINK) $(dhcrelay_OBJECTS) $(dhcrelay_LDADD) $(LIBS)
+#dhcrelay$(EXEEXT): $(dhcrelay_OBJECTS) $(dhcrelay_DEPENDENCIES) 
+#	@rm -f dhcrelay$(EXEEXT)
+#	$(CC) -o $@ $^ $(dhcrelay_LDADD) $(LIBS)
 
 mostlyclean-compile:
 	-rm -f *.$(OBJEXT)
diff -Nur dhcp-4.1.1.orig/server/Makefile.in dhcp-4.1.1/server/Makefile.in
--- dhcp-4.1.1.orig/server/Makefile.in	2012-03-12 19:06:19.460825284 +0800
+++ dhcp-4.1.1/server/Makefile.in	2012-03-12 19:03:08.773450663 +0800
@@ -243,12 +243,12 @@
 
 clean-sbinPROGRAMS:
 	-test -z "$(sbin_PROGRAMS)" || rm -f $(sbin_PROGRAMS)
-#dhcpd$(EXEEXT): $(dhcpd_OBJECTS) $(dhcpd_DEPENDENCIES) 
-#	@rm -f dhcpd$(EXEEXT)
-#	$(LINK) $(dhcpd_OBJECTS) $(dhcpd_LDADD) $(LIBS)
 dhcpd$(EXEEXT): $(dhcpd_OBJECTS) $(dhcpd_DEPENDENCIES) 
 	@rm -f dhcpd$(EXEEXT)
-	$(CC) -o $@ $^ $(dhcpd_LDADD) $(LIBS)
+	$(LINK) $(dhcpd_OBJECTS) $(dhcpd_LDADD) $(LIBS)
+#dhcpd$(EXEEXT): $(dhcpd_OBJECTS) $(dhcpd_DEPENDENCIES) 
+#	@rm -f dhcpd$(EXEEXT)
+#	$(CC) -o $@ $^ $(dhcpd_LDADD) $(LIBS)
 
 mostlyclean-compile:
 	-rm -f *.$(OBJEXT)
